src/compiler/checker.ts

```diff
@@ -19327,8 +19332,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     // there are possible values of the check type that are also possible values of the extends type.
                     // We use a reverse assignability check as it is less expensive than the comparable relationship
                     // and avoids false positives of a non-empty intersection check.
-                    if (checkType.flags & TypeFlags.Any || forConstraint && !(inferredExtendsType.flags & TypeFlags.Never) && someType(getPermissiveInstantiation(inferredExtendsType), t => isTypeAssignableTo(t, getPermissiveInstantiation(checkType)))) {
-                        (extraTypes || (extraTypes = [])).push(instantiateType(getTypeFromTypeNode(root.node.trueType), combinedMapper || mapper));
+                    if (checkType.flags & TypeFlags.Any || distributiveConditionalConstraintIdentityStack.length !== 0 && !(inferredExtendsType.flags & TypeFlags.Never) && someType(getPermissiveInstantiation(inferredExtendsType), t => isTypeAssignableTo(t, getPermissiveInstantiation(checkType)))) {
+                        (extraTypes ??= []).push(instantiateType(getTypeFromTypeNode(root.node.trueType), combinedMapper || mapper));

```

Comment:
this is the core of the fix, previously only the first level of nested conditional type (in non tail-recursive scenarios) was including the instantiated true type in its `extraTypes` (the reasoning behind that is outlined in the comment above and in the https://github.com/microsoft/TypeScript/pull/56004 's description)

So now we are including it at all nested levels. In other words - previously the `forConstraint` was only passed to the direct call of `getConditionalTypeInstantiation` from `getConstraintOfDistributiveConditionalType` but now this contextual~ information is used for all nested calls.

I've used a global variable to track this as it's not really feasible today to thread this through `instantiateType` here.

Reply:
Why is it not feasible to thread `forConstraint` through `instantiateType`?

Reply:
Conditional types could appear just at any deeper level of the instantiated type. It's doable to introduce some kind of `InstantiationFlags` and pass it down to all called functions etc but it felt more complicated to attempt doing it. It would also be fairly easy to forget to pass that down somewhere.